crates/holochain/src/sweettest/sweet_consistency.rs

```diff
@@ -32,60 +32,174 @@ impl DurationOrSeconds {
+    action_seq: u32,
+    author: AgentPubKey,
+    when_integrated: Option<Timestamp>,
 }
 
-/// Wait for all cells to reach consistency,
-/// with the option to specify that some cells are offline.
-///
-/// Cells paired with a `false` value will have their authored ops counted towards the total,
-/// but not their integrated ops (since they are not online to integrate things).
-/// This is useful for tests where nodes go offline.
-#[cfg_attr(feature = "instrument", tracing::instrument(skip_all))]
-pub async fn await_consistency_advanced<'a, I: IntoIterator<Item = (&'a SweetCell, bool)>>(
-    timeout: impl Into<DurationOrSeconds>,
-    conditions: impl Into<ConsistencyConditions>,
-    all_cells: I,
-) -> ConsistencyResult {
-    #[allow(clippy::type_complexity)]
-    let all_cell_dbs: Vec<(
-        AgentPubKey,
-        DbRead<DbKindAuthored>,
-        Option<DbRead<DbKindDht>>,
-    )> = all_cells
-        .into_iter()
-        .map(|(c, online)| {
-            (
-                c.agent_pubkey().clone(),
-                c.authored_db().clone().into(),
-                online.then(|| c.dht_db().clone().into()),
-            )
-        })
-        .collect();
-    let all_cell_dbs: Vec<_> = all_cell_dbs
-        .iter()
-        .map(|c| (&c.0, &c.1, c.2.as_ref()))
-        .collect();
-    wait_for_integration_diff(
-        &all_cell_dbs[..],
-        timeout.into().into_duration(),
-        conditions.into(),
-    )
-    .await
+/// Wait for all cell envs to reach consistency, meaning that every op
+/// published by every cell has been integrated by every node.
+async fn await_op_integration(
+    cells: &[(&AgentPubKey, &impl ReadAccess<DbKindDht>)],
+    timeout: Duration,
+) -> Result<(), String> {
+    let start = Instant::now();
+    // Declare op hash lists here so they can be accessed for reporting after timeout.
+    let mut rows_per_db = Vec::new();
+    let result = tokio::time::timeout(timeout, async {
+        loop {
+            tokio::time::sleep(Duration::from_millis(500)).await;
+            // Create query for each DHT DB.
+            let queries = cells.iter().map(|(_, dht_db)| {
+                dht_db.read_async(|txn| {
+                    let mut stmt = txn
+                        .prepare(
+                            "\
+                            SELECT DhtOp.hash, DhtOp.type, DhtOp.when_integrated, Action.seq, Action.author
+                            FROM DhtOp
+                            JOIN Action ON DhtOp.action_hash = Action.hash",
+                        )
+                        .unwrap();
+                    let mut rows = stmt.query([]).unwrap();
+                    let mut values = Vec::new();
+                    while let Some(row) = rows.next().unwrap() {
+                        let hash = row.get_unwrap::<_, DhtOpHash>(0);
+                        let op_type = row.get_unwrap::<_, DhtOpType>(1);
+                        let when_integrated = row.get_unwrap::<_, Option<Timestamp>>(2);
```

Comment:
Read my next comment first, so if this doesn't make sense ignore it.

Should we `continue` the loop up here if `when_integrated` is None? Since (I think) we're trying to do the same thing later after more work?

Reply:
Nice catch! Yes, that continue was ineffective. continues aren't allowed inside of closure or functions, so I can't put that here. It's not much more work, because these futures defined here are executed a little further down and then already iterated over. I've applied the loop label in the other position.